File: packages/sdk-python/arcenciel/oauth.py
"""OAuth Authorization Code + PKCE helpers for public and confidential clients."""

from __future__ import annotations

import base64
import hashlib
import secrets
from dataclasses import dataclass
from typing import Any
from urllib.parse import parse_qs, urlencode, urlparse

import httpx

from arcenciel.errors import ArcEnCielError


@dataclass(frozen=True, slots=True)
class OAuthAuthorization:
    """Short-lived local state required to validate and exchange an OAuth callback."""

    url: str
    state: str
    code_verifier: str
    client_id: str
    redirect_uri: str
    issuer: str


@dataclass(frozen=True, slots=True)
class OAuthTokenSet:
    """Opaque OAuth credentials returned by an authorization or refresh exchange."""

    access_token: str
    token_type: str
    expires_in: int
    scope: str
    refresh_token: str | None = None


class OAuthFlow:
    """Safe sync/async OAuth flow with state, issuer, PKCE, and rotation handling."""

    def __init__(
        self,
        base_url: str = "https://arcenciel.io",
        timeout: float = 30.0,
        *,
        transport: httpx.BaseTransport | None = None,
        async_transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.transport = transport
        self.async_transport = async_transport

    def create_authorization(
        self,
        *,
        client_id: str,
        redirect_uri: str,
        scopes: list[str],
        state: str | None = None,
    ) -> OAuthAuthorization:
        """Create an authorization URL and local verifier using mandatory PKCE S256."""

        if not client_id or not redirect_uri or not scopes:
            raise ArcEnCielError(
                "client_id, redirect_uri, and at least one scope are required",
                code="INVALID_ARGUMENT",
            )
        oauth_state = state or secrets.token_urlsafe(24)
        if len(oauth_state) < 8:
            raise ArcEnCielError(
                "OAuth state must contain at least 8 characters",
                code="INVALID_ARGUMENT",
            )
        verifier = secrets.token_urlsafe(64)
        challenge = (
            base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest())
            .rstrip(b"=")
            .decode()
        )
        query = urlencode(
            {
                "response_type": "code",
                "client_id": client_id,
                "redirect_uri": redirect_uri,
                "scope": " ".join(scopes),
                "state": oauth_state,
                "code_challenge": challenge,
                "code_challenge_method": "S256",
            }
        )
        return OAuthAuthorization(
            url=f"{self.base_url}/api/oauth/authorize?{query}",
            state=oauth_state,
            code_verifier=verifier,
            client_id=client_id,
            redirect_uri=redirect_uri,
            issuer=self.base_url,
        )

    def metadata(self) -> dict[str, Any]:
        return self._get("/.well-known/oauth-authorization-server")

    async def metadata_async(self) -> dict[str, Any]:
        return await self._get_async("/.well-known/oauth-authorization-server")

    def scopes(self) -> list[dict[str, Any]]:
        payload = self._get("/api/oauth/scopes")
        return list(payload.get("data", []))

    async def scopes_async(self) -> list[dict[str, Any]]:
        payload = await self._get_async("/api/oauth/scopes")
        return list(payload.get("data", []))

    def get_public_client(self, client_id: str) -> dict[str, Any]:
        return dict(self._get(f"/api/oauth/clients/{client_id}/public").get("data", {}))

    async def get_public_client_async(self, client_id: str) -> dict[str, Any]:
        payload = await self._get_async(f"/api/oauth/clients/{client_id}/public")
        return dict(payload.get("data", {}))

    def exchange_callback(
        self,
        *,
        callback_url: str,
        authorization: OAuthAuthorization,
        client_secret: str | None = None,
    ) -> OAuthTokenSet:
        code = self._validate_callback(callback_url, authorization)
        return self._token(
            {
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": authorization.redirect_uri,
                "code_verifier": authorization.code_verifier,
            },
            authorization.client_id,
            client_secret,
        )

    async def exchange_callback_async(
        self,
        *,
        callback_url: str,
        authorization: OAuthAuthorization,
        client_secret: str | None = None,
    ) -> OAuthTokenSet:
        code = self._validate_callback(callback_url, authorization)
        return await self._token_async(
            {
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": authorization.redirect_uri,
                "code_verifier": authorization.code_verifier,
            },
            authorization.client_id,
            client_secret,
        )

    def refresh(
        self,
        *,
        client_id: str,
        refresh_token: str,
        client_secret: str | None = None,
    ) -> OAuthTokenSet:
        """Rotate a refresh token once; replace the previous token atomically."""

        return self._token(
            {"grant_type": "refresh_token", "refresh_token": refresh_token},
            client_id,
            client_secret,
        )

    async def refresh_async(
        self,
        *,
        client_id: str,
        refresh_token: str,
        client_secret: str | None = None,
    ) -> OAuthTokenSet:
        return await self._token_async(
            {"grant_type": "refresh_token", "refresh_token": refresh_token},
            client_id,
            client_secret,
        )

    def revoke(
        self,
        *,
        client_id: str,
        token: str,
        client_secret: str | None = None,
    ) -> None:
        self._form("/api/oauth/revoke", {"token": token}, client_id, client_secret, False)

    async def revoke_async(
        self,
        *,
        client_id: str,
        token: str,
        client_secret: str | None = None,
    ) -> None:
        await self._form_async(
            "/api/oauth/revoke", {"token": token}, client_id, client_secret, False
        )

    def introspect(
        self,
        *,
        client_id: str,
        client_secret: str,
        token: str,
    ) -> dict[str, Any]:
        return self._form("/api/oauth/introspect", {"token": token}, client_id, client_secret, True)

    async def introspect_async(
        self,
        *,
        client_id: str,
        client_secret: str,
        token: str,
    ) -> dict[str, Any]:
        return await self._form_async(
            "/api/oauth/introspect", {"token": token}, client_id, client_secret, True
        )

    def _validate_callback(self, callback_url: str, authorization: OAuthAuthorization) -> str:
        query = parse_qs(urlparse(callback_url).query)
        if "error" in query:
            code = query["error"][0]
            message = query.get("error_description", [code])[0]
            raise ArcEnCielError(message, code=code)
        if query.get("state", [None])[0] != authorization.state:
            raise ArcEnCielError(
                "OAuth callback state does not match the authorization request",
                code="OAUTH_STATE_MISMATCH",
            )
        if query.get("iss", [None])[0] != authorization.issuer:
            raise ArcEnCielError(
                "OAuth callback issuer does not match Arc en Ciel",
                code="OAUTH_ISSUER_MISMATCH",
            )
        authorization_code = query.get("code", [None])[0]
        if not authorization_code:
            raise ArcEnCielError(
                "OAuth callback does not contain an authorization code",
                code="INVALID_CALLBACK",
            )
        return authorization_code

    def _get(self, path: str) -> dict[str, Any]:
        with httpx.Client(
            base_url=self.base_url, timeout=self.timeout, transport=self.transport
        ) as client:
            return self._response(client.get(path))

    async def _get_async(self, path: str) -> dict[str, Any]:
        async with httpx.AsyncClient(
            base_url=self.base_url, timeout=self.timeout, transport=self.async_transport
        ) as client:
            return self._response(await client.get(path))

    def _token(
        self,
        fields: dict[str, str],
        client_id: str,
        client_secret: str | None,
    ) -> OAuthTokenSet:
        return self._token_set(
            self._form("/api/oauth/token", fields, client_id, client_secret, True)
        )

    async def _token_async(
        self,
        fields: dict[str, str],
        client_id: str,
        client_secret: str | None,
    ) -> OAuthTokenSet:
        payload = await self._form_async("/api/oauth/token", fields, client_id, client_secret, True)
        return self._token_set(payload)

    def _form(
        self,
        path: str,
        fields: dict[str, str],
        client_id: str,
        client_secret: str | None,
        expect_json: bool,
    ) -> dict[str, Any]:
        body, auth = self._form_auth(fields, client_id, client_secret)
        with httpx.Client(
            base_url=self.base_url, timeout=self.timeout, transport=self.transport
        ) as client:
            response = (
                client.post(path, data=body, auth=auth)
                if auth is not None
                else client.post(path, data=body)
            )
            return self._response(response) if expect_json else self._empty_response(response)

    async def _form_async(
        self,
        path: str,
        fields: dict[str, str],
        client_id: str,
        client_secret: str | None,
        expect_json: bool,
    ) -> dict[str, Any]:
        body, auth = self._form_auth(fields, client_id, client_secret)
        async with httpx.AsyncClient(
            base_url=self.base_url, timeout=self.timeout, transport=self.async_transport
        ) as client:
            response = (
                await client.post(path, data=body, auth=auth)
                if auth is not None
                else await client.post(path, data=body)
            )
            return self._response(response) if expect_json else self._empty_response(response)

    @staticmethod
    def _form_auth(
        fields: dict[str, str], client_id: str, client_secret: str | None
    ) -> tuple[dict[str, str], tuple[str, str] | None]:
        body = dict(fields)
        if client_secret is None:
            body["client_id"] = client_id
            return body, None
        return body, (client_id, client_secret)

    @staticmethod
    def _empty_response(response: httpx.Response) -> dict[str, Any]:
        OAuthFlow._raise_for_status(response)
        return {}

    @staticmethod
    def _response(response: httpx.Response) -> dict[str, Any]:
        OAuthFlow._raise_for_status(response)
        payload = response.json()
        if not isinstance(payload, dict):
            raise ArcEnCielError(
                "OAuth endpoint returned an invalid response", code="INVALID_RESPONSE"
            )
        return payload

    @staticmethod
    def _raise_for_status(response: httpx.Response) -> None:
        if response.is_success:
            return
        try:
            payload = response.json()
        except ValueError:
            payload = {}
        code = payload.get("error") if isinstance(payload, dict) else None
        message = payload.get("error_description") if isinstance(payload, dict) else None
        raise ArcEnCielError(
            str(message or code or f"OAuth request failed ({response.status_code})"),
            status=response.status_code,
            code=code if isinstance(code, str) else None,
            request_id=response.headers.get("x-request-id"),
            headers=response.headers,
        )

    @staticmethod
    def _token_set(payload: dict[str, Any]) -> OAuthTokenSet:
        return OAuthTokenSet(
            access_token=str(payload.get("access_token", "")),
            token_type=str(payload.get("token_type", "")),
            expires_in=int(payload.get("expires_in", 0)),
            scope=str(payload.get("scope", "")),
            refresh_token=(
                payload["refresh_token"] if isinstance(payload.get("refresh_token"), str) else None
            ),
        )

File: packages/sdk-typescript/src/oauth.ts
import { ArcEnCielError } from './errors.js'
import type { FetchAPI } from './runtime.js'

export interface OAuthAuthorization {
  url: string
  state: string
  codeVerifier: string
  clientId: string
  redirectUri: string
  issuer: string
}

export interface CreateOAuthAuthorizationOptions {
  clientId: string
  redirectUri: string
  scopes: readonly string[]
  state?: string
}

export interface OAuthClientCredentials {
  clientId: string
  clientSecret?: string
}

export interface OAuthTokenSet {
  accessToken: string
  tokenType: string
  expiresIn: number
  scope: string
  refreshToken?: string
}

export interface OAuthTokenIntrospection {
  active: boolean
  clientId?: string
  username?: string
  scope?: string
  tokenType?: string
  exp?: number
  iat?: number
  sub?: string
  iss?: string
}

const base64Url = (bytes: Uint8Array): string => {
  let binary = ''
  for (const value of bytes) binary += String.fromCharCode(value)
  return globalThis.btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
}

const randomUrlSafe = (bytes: number): string => {
  if (!globalThis.crypto?.getRandomValues) {
    throw new ArcEnCielError('Web Crypto random generation is required for OAuth state and PKCE', { code: 'CRYPTO_UNAVAILABLE' })
  }
  return base64Url(globalThis.crypto.getRandomValues(new Uint8Array(bytes)))
}

const basicAuthorization = (clientId: string, clientSecret: string): string => {
  const encoded = new TextEncoder().encode(`${clientId}:${clientSecret}`)
  let binary = ''
  for (const value of encoded) binary += String.fromCharCode(value)
  return `Basic ${globalThis.btoa(binary)}`
}

const normalizeTokenSet = (payload: Record<string, unknown>): OAuthTokenSet => ({
  accessToken: String(payload.access_token ?? ''),
  tokenType: String(payload.token_type ?? ''),
  expiresIn: Number(payload.expires_in),
  scope: String(payload.scope ?? ''),
  ...(typeof payload.refresh_token === 'string' ? { refreshToken: payload.refresh_token } : {}),
})

export class OAuthFlow {
  constructor(
    private readonly baseUrl: string,
    private readonly fetchApi: FetchAPI
  ) {}

  async createAuthorization(options: CreateOAuthAuthorizationOptions): Promise<OAuthAuthorization> {
    if (!options.clientId || !options.redirectUri || options.scopes.length === 0) {
      throw new ArcEnCielError('clientId, redirectUri, and at least one scope are required', { code: 'INVALID_ARGUMENT' })
    }
    const issuer = this.baseUrl
    const state = options.state ?? randomUrlSafe(24)
    if (state.length < 8) throw new ArcEnCielError('OAuth state must contain at least 8 characters', { code: 'INVALID_ARGUMENT' })
    const codeVerifier = randomUrlSafe(64)
    const digest = await globalThis.crypto.subtle.digest('SHA-256', new TextEncoder().encode(codeVerifier))
    const query = new URLSearchParams({
      response_type: 'code',
      client_id: options.clientId,
      redirect_uri: options.redirectUri,
      scope: options.scopes.join(' '),
      state,
      code_challenge: base64Url(new Uint8Array(digest)),
      code_challenge_method: 'S256',
    })
    return {
      url: `${issuer}/api/oauth/authorize?${query.toString()}`,
      state,
      codeVerifier,
      clientId: options.clientId,
      redirectUri: options.redirectUri,
      issuer,
    }
  }

  async exchangeCallback(options: {
    callbackUrl: string
    authorization: OAuthAuthorization
    clientSecret?: string
  }): Promise<OAuthTokenSet> {
    const callback = new URL(options.callbackUrl)
    const protocolError = callback.searchParams.get('error')
    if (protocolError) {
      throw new ArcEnCielError(callback.searchParams.get('error_description') ?? protocolError, { code: protocolError })
    }
    if (callback.searchParams.get('state') !== options.authorization.state) {
      throw new ArcEnCielError('OAuth callback state does not match the authorization request', { code: 'OAUTH_STATE_MISMATCH' })
    }
    if (callback.searchParams.get('iss') !== options.authorization.issuer) {
      throw new ArcEnCielError('OAuth callback issuer does not match Arc en Ciel', { code: 'OAUTH_ISSUER_MISMATCH' })
    }
    const code = callback.searchParams.get('code')
    if (!code) throw new ArcEnCielError('OAuth callback does not contain an authorization code', { code: 'INVALID_CALLBACK' })
    return this.token(
      {
        grant_type: 'authorization_code',
        code,
        redirect_uri: options.authorization.redirectUri,
        code_verifier: options.authorization.codeVerifier,
      },
      { clientId: options.authorization.clientId, clientSecret: options.clientSecret }
    )
  }

  refresh(options: OAuthClientCredentials & { refreshToken: string }): Promise<OAuthTokenSet> {
    return this.token({ grant_type: 'refresh_token', refresh_token: options.refreshToken }, options)
  }

  async revoke(options: OAuthClientCredentials & { token: string }): Promise<void> {
    await this.formRequest('/api/oauth/revoke', { token: options.token }, options, false)
  }

  async introspect(options: OAuthClientCredentials & { clientSecret: string; token: string }): Promise<OAuthTokenIntrospection> {
    const payload = await this.formRequest('/api/oauth/introspect', { token: options.token }, options, true)
    return {
      active: payload.active === true,
      ...(typeof payload.client_id === 'string' ? { clientId: payload.client_id } : {}),
      ...(typeof payload.username === 'string' ? { username: payload.username } : {}),
      ...(typeof payload.scope === 'string' ? { scope: payload.scope } : {}),
      ...(typeof payload.token_type === 'string' ? { tokenType: payload.token_type } : {}),
      ...(typeof payload.exp === 'number' ? { exp: payload.exp } : {}),
      ...(typeof payload.iat === 'number' ? { iat: payload.iat } : {}),
      ...(typeof payload.sub === 'string' ? { sub: payload.sub } : {}),
      ...(typeof payload.iss === 'string' ? { iss: payload.iss } : {}),
    }
  }

  private async token(fields: Record<string, string>, credentials: OAuthClientCredentials): Promise<OAuthTokenSet> {
    return normalizeTokenSet(await this.formRequest('/api/oauth/token', fields, credentials, true))
  }

  private async formRequest(
    path: string,
    fields: Record<string, string>,
    credentials: OAuthClientCredentials,
    expectJson: boolean
  ): Promise<Record<string, unknown>> {
    const headers = new Headers({ 'Content-Type': 'application/x-www-form-urlencoded' })
    const body = new URLSearchParams(fields)
    if (credentials.clientSecret) headers.set('Authorization', basicAuthorization(credentials.clientId, credentials.clientSecret))
    else body.set('client_id', credentials.clientId)
    const response = await this.fetchApi(`${this.baseUrl}${path}`, { method: 'POST', headers, body })
    if (!response.ok) {
      let payload: Record<string, unknown> = {}
      try {
        payload = (await response.clone().json()) as Record<string, unknown>
      } catch {
        // OAuth permits empty error responses from intermediaries.
      }
      throw new ArcEnCielError(String(payload.error_description ?? payload.error ?? `OAuth request failed (${response.status})`), {
        status: response.status,
        code: typeof payload.error === 'string' ? payload.error : undefined,
        requestId: response.headers.get('x-request-id') ?? undefined,
        headers: response.headers,
      })
    }
    if (!expectJson) return {}
    return (await response.json()) as Record<string, unknown>
  }
}
